### DISCUSSION TITLE
Regarding Micrometer Extension

### DISCUSSION THREAD

**Discussion by raushan606** [quarkusio/quarkus#43881] [Q&A] [answered]
Hey,
I require to build an extension, which gives me memory metrics (Heap Usage) and complies with Microprofile Metrics Api 5.0.0
I know those metrics are there, but I require a specific name.

Now, I thought of two approaches here:
First, Rewrite the micrometer extension, and remove unnecessary code (Which I don't need). My only concern is if anyone adds my custom extension, and if a user hits `http://localhost:8080/metrics` then it will give heap usage memory metric.

Second, I'm trying right now and got stuck with it. Code below:
```
@Recorder
public class MetricsRecorder {

    private static final String MEMORY_COMMITTED_HEAP = "memory.usedHeap";
    final String TAG_METRIC_MP_SCOPE_NAME = "mp_scope";
    final String BASE_SCOPE = "base";

    MeterRegistry micrometerRegistry = Metrics.globalRegistry;

    public void baseMemoryMetrics() {
        MemoryMXBean memoryMXBean = ManagementFactory.getMemoryMXBean();
        Gauge.builder(MEMORY_COMMITTED_HEAP, () -> memoryMXBean.getHeapMemoryUsage().getUsed())
                .baseUnit(BaseUnits.BYTES)
                .description("Displays the amount of used heap memory in bytes.")
                .tags(TAG_METRIC_MP_SCOPE_NAME, BASE_SCOPE)
                .register(micrometerRegistry);
    }

}
```

This is in the runtime module, The question is what will be in the buildstep in the deployment module which will use this recorder to give me heap metrics or whatever I wrote in `baseMemoryMetrics()` method.

[a comment by an automated account is not shown]

**Reply by brunobat** [reply]
Microprofile Metrics is considered deprecated in Quarkus and not recommended. The `quarkus-micrometer` extension should be used instead.
That extension already produces jvm memory metrics. You don't need to implement that yourself. 
Just add the dependency and activate the jvm metrics by setting in `application.properties` the following property:
`quarkus.micrometer.binder.jvm=true`
Please check the Micrometer guide here: https://quarkus.io/guides/telemetry-micrometer

**Reply by raushan606** [reply]
Thanks for the quick reply @brunobat 
Yeah, I'm aware of it but I require specific naming for it.
For eg.:

```
# TYPE memory_usedNonHeap_bytes gauge
# HELP memory_usedNonHeap_bytes Displays the amount of used memory in bytes.
memory_usedNonHeap_bytes{mp_scope="base"} 1.41664872E8
```

Is there a way to do it with a micrometer?

**Reply by ebullient** [reply]
Use a MeterFilter to change the name: 
- https://quarkus.io/guides/telemetry-micrometer#meter-filter
- https://docs.micrometer.io/micrometer/reference/concepts/meter-filters.html

**Reply by raushan606** [reply, marked as answer]
Thank you for your answers. I resolved my issue. I use `BuildProducer<MetricsFactoryConsumerBuildItem>` in buildstep and add metrics in the recorder by overriding the `accept()` method.

```
    public Consumer<MetricsFactory> registerMetrics() {
        return new Consumer<MetricsFactory>() {
            @Override
            public void accept(MetricsFactory metricsFactory) {
                usedHeap(metricsFactory);
                usedNonHeap(metricsFactory);
                committedHeap(metricsFactory);
                committedNonHeap(metricsFactory);
                maxHeap(metricsFactory);
                maxNonHeap(metricsFactory);
            }
        };
    }
```
